File: services/thoth/src/datamuse.rs
use std::{collections::HashMap, fmt::Display};

use anyhow::Result;
use serde::{Deserialize, Serialize};
use url_search_params::build_url_search_params;

pub type DatamuseResult = Vec<DatamuseHit>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatamuseHit {
    pub word: String,
    pub score: i32,
}

#[derive(Debug)]
pub enum DatamuseQuery {
    Adjective,
    CloseRhyme,
    Holonym,
    Homophone,
    Hyponym,
    MatchWord,
    Noun,
    Rhyme,
    MeansLike,
    SpelledLike,
	SoundsLike,
	Triggers,
}

impl Display for DatamuseQuery {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let word = match self {
            Self::Adjective => "rel_jjb",
            Self::CloseRhyme => "rel_nry",
            Self::Holonym => "rel_com",
            Self::Homophone => "rel_hom",
            Self::Hyponym => "rel_gen",
            Self::MatchWord => "sp",
            Self::Noun => "rel_jja",
            Self::Rhyme => "rel_rhy",
            Self::MeansLike => "ml",
            Self::SpelledLike => "sp",
			Self::SoundsLike => "sl",
			Self::Triggers => "rel_trg"
        };

        f.write_str(word)
    }
}

/// performs a query on the datamuse api
pub async fn fetch_datamuse(query: DatamuseQuery, word: &String) -> Result<DatamuseResult> {
    let url = format!("https://api.datamuse.com/words?{}={}", query, word);

    let response = reqwest::get(url).await?.json::<DatamuseResult>().await?;

    Ok(response)
}

/// performs a query on the datamuse api with a spelled-like paramater
pub async fn fetch_datamuse_raw(query: HashMap<String, String>) -> Result<DatamuseResult> {
    let qs = build_url_search_params(query);
    let url = format!("https://api.datamuse.com/words?{qs}");

    let response = reqwest::get(url).await?.json::<DatamuseResult>().await?;

    Ok(response)
}

/// shortens an array to a max length
pub fn trim_array(mut array: Vec<String>, max_len: usize) -> Vec<String> {
    if array.len() > max_len {
        let len = array.len() - max_len;
        array.truncate(max_len);
        // new_array.push(len as T);
        let more = format!("{} more...", len);
        array.push(more);

        return array;
    }

    array.to_vec()
}

File: src/util/datamuse.ts
import { URLSearchParams } from 'node:url';

export type DatamuseHit = {
	score: number;
	word: string;
};

export type DatamuseResult = DatamuseHit[];

export enum DatamuseQuery {
	Adjective = 'rel_jjb',
	CloseRhyme = 'rel_nry',
	Holonym = 'rel_com',
	Homophone = 'rel_hom',
	Hyponym = 'rel_gen',
	MeansLike = 'ml',
	Noun = 'rel_jja',
	Rhyme = 'rel_rhy',
	SoundsLike = 'sl',
	SpelledLike = 'sp',
	ThatFollows = 'lc',
	Triggers = 'rel_trg',
}

export async function fetchDatamuse(query: DatamuseQuery, word: string): Promise<DatamuseResult> {
	const params = new URLSearchParams();
	params.append(query, word);

	return fetchDatamuseRaw(params);
}

export async function fetchDatamuseRaw(query: URLSearchParams): Promise<DatamuseResult> {
	const url = `https://api.datamuse.com/words?${query}`;

	const response = await fetch(url, {
		headers: {
			'user-agent': 'Thoth (github.com/Fyko/Thoth)',
			accept: 'application/json',
		},
	});

	return (await response.json()) as DatamuseResult;
}
